### ISSUE TITLE
maxSurge for node draining or how to meet availability requirements when draining nodes by adding pods

### ISSUE THREAD

**Issue by txomon** [kubernetes/kubernetes#114877]
### What would you like to be added?

A way to drain nodes by adding more pods elsewhere to meet PodDisruptionBudgets.

### Why is this needed?

Currently, when there is a Deployment, it can be configured to have a `maxSurge` to avoid going under the amount of replicas the deployment requires while allowing for a new release to be rolled out. This parameter allows adding extra pods before subtracting the old ones so that the "replicas" number required is always met as a minimum,

This feature (to my knowledge) is only available when releasing new versions of an application, however when draining nodes this would be extremely useful.

Usual cluster maintenance is done by adding new nodes before removing old ones. This means all the pods in the node need to be evicted and there is usually space for one more of each of the old node in the new node. Current solutions such as the PodDisruptionBudget or Eviction API are trying to make sure that substracting pods from the current amount don´t break anything, however the possibility of temporarily having one extra pod of each deployment is not contemplated at the moment.

This request is asking for the ability to use a surplus of pods to meet all constraints for safe eviction.

Some side notes to stress the importance. Although when operating evictions on large workloads lack of PDBs or PDBs with minAvailable/maxUnavailable settings work fine. When moving deployments with 1 replicas or HPA controlled deployments that are currently scaled down enough the problem is aggravated and can only be solved through a few inefficient means, which is acerbated if node maintenance is done automatically (such as GKE, and other cloud services)

Just in case, this is a limitation that should only be counted against Deployments with `strategy.type=RollingUpdate`.

Ways to deal with this situation currently:

1. Have a `minReplicas`/`replicas` to `>1`, and a PDB with `maxUnavailable=1` when it's known that the autoscaler if in use, it's usually scaled on the lower end. Pros: There is no downtime, Cons: Waste of resources.
2. Do nothing, and deal with eventual downtimes. Pros: No waste of resources, Cons: There is downtime in the deployment

**Comment by james-callahan** [comment]
This is still a huge obstacle for gracefully replacing nodes.
Is anyone trying to take this on?

**Comment by txomon** [comment]
I don't think so, it seems like it isn't gathering much attention, I'm not sure if the working group is even aware of it...

**Comment by neolit123** [comment]
/remove-wg reliability
/remove-sig scalability
/remove-sig cluster-lifecycle
/remove-sig autoscaling
/sig node

sig apps is the primary owner of this FR due to the maxSurge feature. you can join their zoom meeting and present it:
https://github.com/kubernetes/community/tree/master/sig-apps
sig node can participate with respect to node draining (as the title indicates).

other sigs / wg should only be added once maintainers have a look at this ticket.

**Comment by soltysh** [comment]
It's important to understand that any controller (deployment, or any other workload) and disruption are two distinct mechanism having different roles, although their  functionality is complementary when we're talking about ensuring high availability.
I'd like to stress here the high availability factor, which is at the front and center when talking about PDBs. I will admit it's challenging to have any further discussion without that prerequisite fullfilled. 
In a similar vein, when talking about HPAs, there's an option for minimum number of replicas, which ensure the application always maintains the HA pre-reqs. 

Moving on to the subject of configuring a rolling update (in case of a deployment), versus responding to external disruptions (in this case, PDB being an external actor is equal to invoking `kubectl delete` by the user). The reason we have
the ability to have a detailed rollout is coming from the fact that it is a process fully operated by the owning controller. Whereas in all other situations (ie. any external errors or disruptions), the controller's goal is to reach the desired state as quickly as possible. 

Speaking from my own experience, we've had multiple instances of problems when PDBs were blocking the upgrades, due to users setting the allowed replicas down to bare minimum. We solved it by [adding an alert](https://github.com/openshift/cluster-kube-controller-manager-operator/blob/d95b0c25ba55c4ef8e09e56461562ee60b22d51c/manifests/0000_90_kube-controller-manager-operator_05_alerts.yaml#L25-L44) which looks at PDBs and notifies administrators in those cases. Since then we haven't seen any problems, or cluster administrators were aware that problems might popup during upgrade and were able to solve these problems even before initiating the upgrades. 

Having said all of the above, [SIG Apps](https://github.com/kubernetes/community/tree/master/sig-apps) meeting happens on every other Monday, next occurrence is planned for September 18. If you're interested, I'd be happy to hear more about your use cases and the problems you're struggling with.

**Comment by txomon** [comment]
Hello Maciej,

I will do my best to join the monday call.

Just for other people reading the thread, my use case are all the
miscellaneous apps that are needed to run some services on top of
kubernetes (such as knative admission controller, external-dns, etc.), as
well as any other app that just doesn't need scaling.

These apps have PDBs that make sure that there is always at least a single
instance running (hence it makes sense that the minimum availability of the
PDB is 1), however running more than one instance at a time would be a
waste of resources.

I understand there is ownership and interactions between the different
controllers, and some of my ideas were revolving around creating a new
attribute, however it was brought to my attention that the `maxSurge` field
perfectly describes the situation, "I'm okay with having up to X instances
for a while".

Regarding the PDB blocking upgrade issue, I had those too, however because
we are running in GKE and nodes get recycled in a regular basis, we can't
have a cluster operator be waiting for a node maintenance to happen (nor we
want to).

I hope I was able to give enough context,

Cheers, Javier

**Comment by james-callahan** [comment]
Likewise: usually when I run into this its when a low availability requirement cluster is running cluster-critical services (e.g. an admission controller) where it is a complete waste of resources to run multiple replicas, but the need to drain without manual interaction (e.g. in response to a spot instance removal) is important.

I'm unable to join the SIG apps call, as the meeting time is incompatible with Australian timezones.

**Comment by james-callahan** [comment]
/remove-lifecycle stale

Where can I follow any ongoing discussion?

**Comment by paulgmiller** [comment]
Still need to read this [proposed KEP]( https://github.com/kubernetes/enhancements/pull/4213/files ) from above but generally trying to deal with similar situations where a deployment surge might unblock evictions and let node drains proceed
https://github.com/paulgmiller/k8s-pdb-autoscaler

But I see that as more of a proof of concept. It gets better if we had something like https://github.com/kubernetes/kubernetes/issues/128815
To tell us about blocked evictions (the webhook solution I have today works but who wants a webhook)

If this was useful other operators could use it and maybe eventually replicasets could bubble up to deployment that they were under eviction pressure then the deployment could decide if they want to use surge or not to try and alleviate that pressure.

**Comment by atiratree** [comment]
> Still need to read this [proposed KEP](https://github.com/kubernetes/enhancements/pull/4213/files)

This is not the relevant KEP anymore, it has been spun off into https://github.com/kubernetes/enhancements/pull/4565 which is trying to address this issue.

**Comment by ePaul** [comment]
@atiratree What is the status of this?
1.36 was already released a while ago – did this make it in?

Also, what is then left to do for this feature?

**Comment by kylan11** [comment]
I made a tool to solve this exact issue. https://github.com/kylan11/understudy
